crates/autopilot/src/run_loop.rs

```diff
@@ -405,21 +405,44 @@ impl RunLoop {
 
     /// Execute the solver's solution. Returns Ok when the corresponding
     /// transaction has been mined.
-    async fn settle(&self, driver: &infra::Driver, solved: &Solution) -> Result<(), SettleError> {
+    async fn settle(
+        &self,
+        driver: &infra::Driver,
+        solved: &Solution,
+        auction_id: i64,
+    ) -> Result<(), SettleError> {
         let order_ids = solved.order_ids().copied().collect();
         self.persistence
             .store_order_events(order_ids, OrderEventLabel::Executing);
 
+        let tag = auction_id.to_be_bytes();
         let request = settle::Request {
             solution_id: solved.id,
         };
-
-        let tx_hash = driver
-            .settle(&request, self.max_settlement_transaction_wait)
-            .await
-            .map_err(SettleError::Failure)?
-            .tx_hash;
-
+        let settlement_tx_wait_task =
+            self.wait_for_settlement_transaction(&tag, self.submission_deadline);
+        let driver_settle_task = driver.settle(&request, self.max_settlement_transaction_wait);
+        // Wait for either the settlement transaction to be mined or the driver returned
+        // a result.
+        let tx_hash = match futures::future::select(
+            Box::pin(settlement_tx_wait_task),
+            Box::pin(driver_settle_task),
```

Comment:
nit: putting these tasks into local variables seems like it just makes the code more verbose since the function names are already very descriptive
```suggestion
            Box::pin(self.wait_for_settlement_transaction(&tag, self.submission_deadline)),
            Box::pin(driver.settle(&request, self.max_settlement_transaction_wait)),
```